src/bin/server/worker.rs

```diff
@@ -341,3 +332,70 @@ impl Worker {
         Ok(())
     }
 }
+
+/// Keeps track of the recently seen IDs for both the failed build reports and the learned jobs.
+/// Only the most recent IDs are stored, to avoid growing the memory usage endlessly.
+///
+/// Internally this uses both an HashSet to provide fast lookups and a VecDeque to know which old
+/// jobs needs to be removed.
+struct RecentlySeen<T: Clone + Eq + Hash> {
+    size: usize,
+    lookup: HashSet<T>,
+    removal: VecDeque<T>,
+}
+
+impl<T: Clone + Eq + Hash> RecentlySeen<T> {
+    fn new(size: usize) -> Self {
+        Self {
+            size,
+            lookup: HashSet::new(),
+            removal: VecDeque::new(),
```

Comment:
nit: it probably makes sense to immediately with_capacity these with size, though at our scale it certainly doesn't matter